File: algo/92. Reverse Linked List II.js
// Linked List

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} m
 * @param {number} n
 * @return {ListNode}
 */
var reverseBetween = function(head, m, n) {
    var dummy = new ListNode(0);
    dummy.next = head;
    var cursor = dummy;
    for (var i = 0; i < m-1; i ++) {
        if (cursor.next === null) {
            break;
        }
        cursor = cursor.next;
    }
    // cursor is used to store the node right before the mutation
    // cursor2 is used to store the node of first mutation
    // cursor2 will become the tail of reversed linked list
    var prev = cursor.next;
    var cursor2 = prev;
    var num_steps = n - m;
    
    curr = prev.next;
    while (curr !== null && num_steps != 0) {
        var temp = curr.next;
        curr.next = prev;
        prev = curr;
        curr = temp;
        num_steps --;
    }
    cursor.next = prev;
    cursor2.next = curr;
    
    return dummy.next;
};

File: round2/92. Reverse Linked List II.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode ReverseBetween(ListNode head, int m, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode cursor = dummy;
        int start = m-1;
        while (start>0) {
            cursor = cursor.next;
            start--;
        }
        ListNode save = cursor.next;
        ListNode prev = cursor.next;
        ListNode curr = cursor.next.next;
        int steps = n-m;
        while (curr != null && steps > 0) {
            ListNode tmp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = tmp;
            steps--;
        }
        cursor.next = prev;
        save.next = curr;
        return dummy.next;
    }
}